### ISSUE TITLE
Spyne generates a different WSDL on each restart

### ISSUE THREAD

**Issue by RazZziel** [arskom/spyne#530]
Every time I restart my service, the WSDL Spyne generates is completely different, which makes all the clients fail to validate replies against the old XSD, or even fail to build against the new WSDL.

For instance, I have declared this type:

    class WhiteListEntry(ComplexModel):
        (...)
        state = Unicode(values=('Inactive', 'Active', 'Cancelled')).customize(nullable=False, min_occurs=1)
        (...)

which is used in these RPCs:

    @rpc(UnsignedInteger32,
         _returns=WhiteListEntry.customize(nullable=False, min_occurs=1),
         _throws=[UnauthenticatedError, ResourceNotFound],
         _soap_port_type="WhiteListService")
    def get_entry(ctx, whitelist_entry_id):
        (...)

    @rpc(UnsignedInteger32.customize(nullable=False, min_occurs=1),
         Client,
         DateTime,
         Integer,
         Array(AccessPermission),
         WhiteListEntry.state.customize(nullable=True, min_occurs=1),
         _returns=UnsignedInteger32.customize(nullable=False, min_occurs=1),
         _throws=[UnauthenticatedError, ResourceNotFound],
         _soap_port_type="WhiteListService")
    def add_entry(ctx, parking_id, client, start, duration, access_permissions, state):
        (...)

    @rpc(UnsignedInteger32,
         DateTimeRange,
         DateTimeRange,
         DateTimeRange,
         DateTimeRange,
         WhiteListEntry.state.customize(nullable=True, min_occurs=1),
         _returns=Array(WhiteListEntry).customize(nullable=False, min_occurs=1),
         _throws=[UnauthenticatedError],
         _soap_port_type="WhiteListService")
    def get_entries(ctx, parking_id, created_range, updated_range, start_range, end_range, state):
        (...)

With that code, each time I restart the service, the type is defined under very different names, which break interoperability with all clients:

```
17:50:50 ~$ ssh operator@25.21.140.124 sudo restart spyne-api >/dev/null && sleep 1 && for i in $(curl -s http://example.com/?wsdl | xmllint --xpath '//*[contains(@name, "stateType")]//@name' -); do echo $i; done | sort
name="add_entry_stateType"
name="add_entry_stateType_stateParentType"
name="add_entry_stateType_stateParentType_stateParentParentType"
name="get_entries_stateType"

17:51:00 ~$ ssh operator@25.21.140.124 sudo restart spyne-api >/dev/null && sleep 1 && for i in $(curl -s http://example.com/?wsdl | xmllint --xpath '//*[contains(@name, "stateType")]//@name' -); do echo $i; done | sort
name="add_entry_stateType"
name="get_entries_stateType"
name="WhiteListEntry_stateType"
name="WhiteListEntry_stateType_stateParentType"

17:51:07 ~$ ssh operator@25.21.140.124 sudo restart spyne-api >/dev/null && sleep 1 && for i in $(curl -s http://example.com/?wsdl | xmllint --xpath '//*[contains(@name, "stateType")]//@name' -); do echo $i; done | sort
name="add_entry_stateType"
name="get_entries_stateType"
name="WhiteListEntry_stateType"
name="WhiteListEntry_stateType_stateParentType"

17:51:14 ~$ ssh operator@25.21.140.124 sudo restart spyne-api >/dev/null && sleep 1 && for i in $(curl -s http://example.com/?wsdl | xmllint --xpath '//*[contains(@name, "stateType")]//@name' -); do echo $i; done | sort
name="add_entry_stateType"
name="get_entries_stateType"
name="WhiteListEntry_stateType"
name="WhiteListEntry_stateType_stateParentType"

17:51:18 ~$ ssh operator@25.21.140.124 sudo restart spyne-api >/dev/null && sleep 1 && for i in $(curl -s http://example.com/?wsdl | xmllint --xpath '//*[contains(@name, "stateType")]//@name' -); do echo $i; done | sort
name="add_entry_stateType"
name="get_entries_stateType"
name="WhiteListEntry_stateType"
name="WhiteListEntry_stateType_stateParentType"

17:51:23 ~$ ssh operator@25.21.140.124 sudo restart spyne-api >/dev/null && sleep 1 && for i in $(curl -s http://example.com/?wsdl | xmllint --xpath '//*[contains(@name, "stateType")]//@name' -); do echo $i; done | sort
name="add_entry_stateType"
name="add_entry_stateType_stateParentType"
name="add_entry_stateType_stateParentType_stateParentParentType"
name="get_entries_stateType"

17:51:30 ~$ ssh operator@25.21.140.124 sudo restart spyne-api >/dev/null && sleep 1 && for i in $(curl -s http://example.com/?wsdl | xmllint --xpath '//*[contains(@name, "stateType")]//@name' -); do echo $i; done | sort
name="add_entry_stateType"
name="get_entries_stateType"
name="WhiteListEntry_stateType"
name="WhiteListEntry_stateType_stateParentType"
```

**Comment by RazZziel** [comment]
Looks like even if I manage to get the types right (by restarting the service until they look like the original types I was using), my client fails because of schema validation.

These are some example errors, I get a different error every time I restart my service:

```System.Web.Services.Protocols.SoapHeaderException: :1:0:ERROR:SCHEMASV:SCHEMAV_ELEMENT_CONTENT: Element '{models}name': This element is not expected. Expected is ( {models}surname ).```

```System.Web.Services.Protocols.SoapHeaderException: :1:0:ERROR:SCHEMASV:SCHEMAV_ELEMENT_CONTENT: Element '{models}surname': This element is not expected. Expected is ( {models}credentials ).```

```System.Web.Services.Protocols.SoapHeaderException: :1:0:ERROR:SCHEMASV:SCHEMAV_ELEMENT_CONTENT: Element '{models}name': This element is not expected. Expected is ( {models}credentials ).```

Maybe it's because of the order of the attributes of a type changes every time I restart my server? Exmaples:
```
<xs:complexType name="Client">
    <xs:sequence>
        <xs:element name="surname" type="xs:string"/>
        <xs:element name="credentials" type="s0:CredentialArray"/>
        <xs:element name="name" type="xs:string" nillable="true"/>
    </xs:sequence>
</xs:complexType>
```
```
<xs:complexType name="Client">
    <xs:sequence>
        <xs:element name="credentials" type="s0:CredentialArray"/>
        <xs:element name="surname" type="xs:string"/>
        <xs:element name="name" type="xs:string" nillable="true"/>
    </xs:sequence>
</xs:complexType>
```
```
<xs:complexType name="Client">
    <xs:sequence>
        <xs:element name="name" type="xs:string" nillable="true"/>
        <xs:element name="credentials" type="s0:CredentialArray"/>
        <xs:element name="surname" type="xs:string"/>
    </xs:sequence>
</xs:complexType>
```

**Comment by RazZziel** [comment]
Ok, looks like it's my bad, seems to be fixable with `declare_order`:
* https://github.com/arskom/spyne/blob/master/doc/source/manual/03_types.rst#complex
* http://spyne.io/docs/2.10/reference/model/complex.html#spyne.model.complex.ComplexModelBase.Attributes.declare_order

**Comment by RazZziel** [comment]
Side-note: the solution proposed in https://github.com/arskom/spyne/blob/master/doc/source/manual/03_types.rst#complex didn't work for me:

```
PredictableComplexModel = ComplexModel.customize(declare_order="name")

class Permission(PredictableComplexModel):
    application = Unicode
    feature = Unicode
```

Because besides getting tons of `seems to be a customized class. They are not supposed to be instantiated.` warnings, I started getting errors even on SRPC parameters I'm not instantiating, or instance because customization breaks truth value testing. For instance:

```
    class Client(ComplexModel):
        name = Unicode.customize(nullable=True, min_occurs=1)
        surname = Unicode.customize(nullable=False, min_occurs=1)

    @rpc(Client)
    def add_entry(ctx, client):
        if client:
            print("Client provided")
```
will correctly detect wether or not the `client` object sent by the SOAP client is null or not, but
```
    PredictableComplexModel = ComplexModel.customize(declare_order="name")

    class Client(PredictableComplexModel):
        name = Unicode.customize(nullable=True, min_occurs=1)
        surname = Unicode.customize(nullable=False, min_occurs=1)

    @rpc(Client)
    def add_entry(ctx, client):
        if client:
            print("Client provided")
```
in this code, `if client:` will **always** evaluate to **false**, so it will never detect a `client` object has been sent by the SOAP client.

I fixed it by monkey-patching Spyne before declaring my models:

```
ComplexModel.Attributes.declare_order = "declared"
```

**Comment by plq** [comment]
Hey,

> I fixed it by monkey-patching Spyne before declaring my models:

``declared_order`` and friends are all ugly hacks on Python 2. The right thing to do in Python 2 is to use `_type_info = [ (...) ]` syntax.

That'll work just as well though.

> Side-note: the solution proposed in https://github.com/arskom/spyne/blob/master/doc/source/manual/03_types.rst#complex didn't work for me:

Fixed by 3125a1f.

> in this code, if client: will always evaluate to false, so it will never detect a client object has been sent by the SOAP client.

This, I don't understand. If you could provide a working example, I can investigate it.

That said, operationless ifs have always been dodgy in dynamically-typed languages like Python. Does it work with ``if client is not None:`` ?

**Comment by RazZziel** [comment]
> declared_order and friends are all ugly hacks on Python 2. The right thing to do in Python 2 is to use _type_info = [ (...) ] syntax.
> 
> That'll work just as well though.

I'm using Python 3.6.0 with Spyne 2.12.14, `declared` and `random` are the only orders available in my `.venv/lib/python3.6/site-packages/spyne/model/complex.py`:
```
def _get_ordered_attributes(cls_name, cls_dict, attrs):
    (...)

    SUPPORTED_ORDERS = ('random', 'declared')
    if (attrs.declare_order is not None and
            not attrs.declare_order in SUPPORTED_ORDERS):

        msg = "The declare_order attribute value %r is invalid in %s"
        raise Exception(msg % (attrs.declare_order, cls_name))
```

> Fixed by 3125a1f.

Nice, thanks! Although it's a bit confusing that you still recommend using `name`, when https://github.com/arskom/spyne/blob/master/spyne/model/complex.py#L395 forbids its use (I tried, got a `"The declare_order attribute value %r is invalid in %s"` error)

> This, I don't understand. If you could provide a working example, I can investigate it.

Hmm, I won't be able for a while, I'm sorry :\

> That said, if without operators have always been dodgy in dynamically-typed languages like Python. Does it work with if client is not None: ?

It works like this for me:
* When `Client` as a subclass of `ComplexModel`
  * When the SOAP client sends a **valid** `client`, in the server's RPC `if client:` evaluates to `true`
  * When the SOAP client sends a **null** `client`, in the server's RPC `if client:` evaluates to `false`
* When `Client` as a subclass of `ComplexModel.customize(declare_order="name")`
  * When the SOAP client sends a **valid** `client`, in the server's RPC `if client:` evaluates to `false` <-- Error
  * When the SOAP client sends a **null** `client`, in the server's RPC `if client:` evaluates to `false`
